File: packages/protocol-schema/src/models/workspace.ts
import { Type, type Static } from "@sinclair/typebox";

export const WorkspaceSchema = Type.Object(
  {
    id: Type.String({ format: "uuid" }),
    name: Type.String(),
    rootPath: Type.String(),
    subPaths: Type.Array(Type.String()),
    providerId: Type.String(),
    createdAt: Type.Integer(),
    lastOpenedAt: Type.Integer(),
  },
  { $id: "Workspace" }
);
export type Workspace = Static<typeof WorkspaceSchema>;

export const WorkspaceCreateInputSchema = Type.Object(
  {
    name: Type.String(),
    rootPath: Type.String(),
    subPaths: Type.Optional(Type.Array(Type.String())),
    providerId: Type.String(),
  },
  { $id: "WorkspaceCreateInput" }
);
export type WorkspaceCreateInput = Static<typeof WorkspaceCreateInputSchema>;

export const FileTreeNodeSchema = Type.Object(
  {
    name: Type.String(),
    path: Type.String(),
    isDirectory: Type.Boolean(),
    size: Type.Optional(Type.Integer()),
    children: Type.Optional(Type.Array(Type.Any())), // Recursive FileTreeNode
  },
  { $id: "FileTreeNode" }
);
export type FileTreeNode = Static<typeof FileTreeNodeSchema>;

File: crates/canywhere-protocol/src/models/workspace.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct Workspace {
    pub id: String,
    pub name: String,
    pub root_path: String,
    pub sub_paths: Vec<String>,
    pub provider_id: String,
    pub created_at: i64,
    pub last_opened_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceCreateInput {
    pub name: String,
    pub root_path: String,
    pub sub_paths: Option<Vec<String>>,
    pub provider_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct FileTreeNode {
    pub name: String,
    pub path: String,
    pub is_directory: bool,
    pub size: Option<i64>,
    pub children: Option<Vec<FileTreeNode>>,
}
